rakelib/user_email_identifier.rake

```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+require 'roo'
+
+desc 'Build a CSV of user email identifiers'
+
+namespace :user_email_identifier, type: :task do
+  task :build_csv, %i[uuid_list] => :environment do |_, args|
+    xlsx = Roo::Excelx.new(args[:uuid_list])
+    xlsx.sheet(0)
+    row_count = xlsx.last_row - 1
+    puts "[UserEmailIdentifier] Building CSV from #{args[:uuid_list]}, processing #{row_count} rows..."
+
+    CSV.open('./tmp/user_email_identifiers.csv', 'w') do |csv|
+      csv << %w[data_type form_type va_gov_submission_created_at va_gov_user_uuid va_gov_credential_email]
+      xlsx.each_row_streaming(offset: 1, pad_cells: true) do |row|
+        next if row[3].nil? || row[3].cell_value.nil?
+
+        original_columns = row[0..3].map { |cell| cell && cell.cell_value.to_s.strip }
+        csp_uuid = original_columns[3]
+        user_verification = UserVerification.find_by(idme_uuid: csp_uuid) ||
+                            UserVerification.find_by(backing_idme_uuid: csp_uuid) ||
+                            UserVerification.find_by(logingov_uuid: csp_uuid)
```

Comment:
You can refactor this to be a single query.

```ruby
user_verification = UserVerification.where(
  "idme_uuid = :uuid OR backing_idme_uuid = :uuid OR logingov_uuid = :uuid",
  uuid: csp_uuid
).first
```

Or like this:

```ruby
user_verification = UserVerification
                      .where(idme_uuid: csp_uuid)
                      .or(UserVerification.where(backing_idme_uuid: csp_uuid))
                      .or(UserVerification.where(logingov_uuid: csp_uuid))
                      .first

Reply:
I think there is still the N+1 problem, though it is less severe with the updated query. I'm trying to think of how we could avoid the N+1 altogether.

Reply:
So, I did some refactoring and this is what I came up with:

```ruby
# frozen_string_literal: true

require 'roo'

namespace :user_email_identifier do
  desc 'Build a CSV of user email identifiers'
  task :build_csv, %i[uuid_list] => :environment do |_, args|
    xlsx = Roo::Excelx.new(args[:uuid_list])
    xlsx.sheet(0)
    row_count = xlsx.last_row - 1
    puts "[UserEmailIdentifier] Building CSV from #{args[:uuid_list]}, processing #{row_count} rows..."

    uuids = xlsx.map.with_index { |row, index| row[3] if index.positive? && row[3] }.compact

    user_verifications = UserVerification.where(idme_uuid: uuids)
                                         .or(UserVerification.where(backing_idme_uuid: uuids))
                                         .or(UserVerification.where(logingov_uuid: uuids))
                                         .includes(:user_credential_email)
                                         .index_by do |uv|
                                           [uv.idme_uuid, uv.backing_idme_uuid, uv.logingov_uuid].compact.first
                                         end

    CSV.open('./tmp/user_email_identifiers.csv', 'w') do |csv|
      csv << %w[data_type form_type va_gov_submission_created_at va_gov_user_uuid va_gov_credential_email]
      xlsx.each_row_streaming(offset: 1, pad_cells: true) do |row|
        next if row[3].nil? || row[3].cell_value.nil?

        original_columns = row[0..3].map { |cell| cell && cell.cell_value.to_s.strip }
        csp_uuid = original_columns[3]
        user_verification = user_verifications[csp_uuid]
        if user_verification
          user_credential_email = user_verification.user_credential_email.credential_email
          csv << (original_columns + [user_credential_email])
        else
          puts "[UserEmailIdentifier] Error: UserVerification not found for uuid: #{csp_uuid}"
          csv << original_columns
        end
      end
    end
    puts '[UserEmailIdentifier] CSV creation successful, file location: tmp/user_email_identifiers.csv'
  end
end
```

It avoids the N+1 by pulling out the UUIDs and querying for the associated user verifications outside of the loop, indexing the results of the query with the CSP UUID from the workbook.